### PR TITLE
backend code

### PR DESCRIPTION
**Pull request by excellencetechnologies**

### CODE REVIEW

File: backend/data_validation/validation.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+async function validateDeviceData(checkBody, validationErrors, body) {
+    try {
+        checkBody('chip_id').notEmpty().withMessage("required city");
+        checkBody('user_id').notEmpty()
+        let errors = validationErrors();
+        return body
+    } catch (err) {
+        res.status(400).json({ error: 1, message: "error during validating the data" });
```
Comment (excellencetechnologies) [inline backend/data_validation/validation.js]: how are you getting res object here?

File: backend/middleware/authentication.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
+var User = require("../model/user");
+var jwt = require("jsonwebtoken");
+var Device = require("../model/device");
+
+module.exports = {
+
+    checkForAlreadyUser: (req, res, next) => {
```
Comment (excellencetechnologies) [inline backend/middleware/authentication.js]: use better name "checkForAlreadyUser"

File: backend/middleware/authentication.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                } else {
+                    res.status(200).json({ error: 1, message: "please provide unique email name " });
+                }
+            }
+        })
+    },
+
+    validateToken: (req, res, next) => {
+        const token = req.headers.token;
+        jwt.verify(token, "thanos", (err, decoded) => {
```
Comment (excellencetechnologies) [inline backend/middleware/authentication.js]: what is thanos?

File: backend/middleware/authentication.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            if (err) {
+                res.status(400).json({ error: 1, message: err.message });
+            } else {
+                req.id = decoded.id;
+                next();
+            }
+        })
+    },
+
+    checkForRepeatAction: (req, res, next) => {
```
Comment (excellencetechnologies) [inline backend/middleware/authentication.js]: what is repeat? didnt get this

File: backend/data_validation/validation.js (hunk 1 above)
Comment (excellencetechnologies) [reply backend/data_validation/validation.js]: sir, this is wrong, i need return error from this function then i will handle that at calling time

File: backend/middleware/authentication.js (hunk 2 above)
Comment (excellencetechnologies) [reply backend/middleware/authentication.js]: ok

File: backend/middleware/authentication.js (hunk 3 above)
Comment (excellencetechnologies) [reply backend/middleware/authentication.js]: sir , actually i just checking the code for working purpose i will make a variable for that

File: backend/middleware/authentication.js (hunk 4 above)
Comment (excellencetechnologies) [reply backend/middleware/authentication.js]: sir , if user already inserted device and then he is again trying to insert then i will check here

File: backend/data_validation/validation.js [hunk 5]
```diff
@@ -0,0 +1,12 @@
+async function validateDeviceData(checkBody, validationErrors, body) {
+    checkBody('chip_id').notEmpty().withMessage("required city");
```
Comment (excellencetechnologies) [inline backend/data_validation/validation.js]: ?

File: backend/middleware/authentication.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+var Device = require("../model/device");
+const key = "thanos";
+
+module.exports = {
+
+    checkForAlreadyRegisterUser: (req, res, next) => {
+        let body = req.body;
+        User.findOne({ email: body.email }, (err, obj) => {
+            if (err) {
+                res.status(500).json({ error: 1, message: "mongo db internel problem while saving the finding the user" });
```
Comment (excellencetechnologies) [inline backend/middleware/authentication.js]: ?

File: backend/routes/user.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+            res.status(500).json({ error: 1, message: "mongodb internel problem while saving the user" });
+        } else {
+            res.status(200).json({ status: 1, message: "successfully created user", date: obj });
+        }
+    });
+});
+
+router.post("/login", (req, res) => {
+    let body = req.body;
+    User.findOne({ email: body.email, password: md5(body.password) }, (err, obj) => {
```
Comment (reactexcel) [inline backend/routes/user.js]: add validation here

File: backend/routes/user.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+var router = express.Router();
+var User = require("../model/user");
+var middleware = require("../middleware/authentication");
+var md5 = require("md5");
+var jwt = require("jsonwebtoken");
+const key = "thanos";
+
+router.post("/register", middleware.checkForAlreadyRegisterUser, (req, res) => {
+    body = req.body;
+    let newUser = new User({ name: body.name, password: md5(body.password), email: body.email });
```
Comment (reactexcel) [inline backend/routes/user.js]: add validation

File: backend/routes/user.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
+var express = require("express");
+var router = express.Router();
+var User = require("../model/user");
+var middleware = require("../middleware/authentication");
+var md5 = require("md5");
+var jwt = require("jsonwebtoken");
+const key = "thanos";
```
Comment (reactexcel) [inline backend/routes/user.js]: use config or envronment file for keys and keys should be random

**Pull request closed without merging**